### ISSUE TITLE
Centralize the threading strategy in substrate-service for improved testing and debuggability

### ISSUE THREAD

**Issue by tomaka** [paritytech/substrate#2536]
*This is a long-term issue for a general refactor of the threading strategy of Substrate.*

Let's have a briefly overview of how Substrate works at the moment:

- We have the client, which, in the case of a light client, holds an Arc to the network (through the on_demand service) and sends messages to the network.
- We have the import queue, which holds an Arc to the client and queries it.
- We have the network, that spawns a background thread, and holds an Arc to the client, the import queue, the transaction pool, and sends messages to them.
- We have grandpa, which holds an Arc to the network, sends messages to it, and receives back gossiping messages through a channel.
- We have the informant, that spawns a background thread, holds an Arc to the client, the network, and the transaction pool, and grabs data from them.
- We have the RPC service, which spawns a background thread and holds Arcs to the client and the network and grabs data from them.
- We have the telemetry, which spawns a background thread, and holds closures to be executed on connection, which includes Arcs to sub-objects.
- And finally we have the service, which initializes everything, and does the additional plugging in that all the other services don't perform, such as notifying the network when the import queue receives something.

This architecture surely is decentralized, but it has quite a few issues:

- Since there are Arcs and threads everywhere, there are also mutexes everywhere. It is quite hard to debug lock contention issues. And in terms of metrics, we generally don't know what's taking the most time.
- There are at least two cyclic dependencies (A contains an Arc to B, and B contains an Arc to A). This leads to some awkward `stop` methods.
- All the APIs that use channels are concurrent, and it's arguably harder and more error-prone to use a concurrent API compared to a basic synchronous API. It is extremely valuable to have APIs that return an error in case of bad input, but if you send your message over a channel you cannot have that feedback.
- It is extremely hard to discover bugs caused by services being out of sync. For example, service A may think that we're connected to a peer, while service B may think that we're not. Considering that everything is asynchronous, we can't really differentiate between a normal situation (the two services are not in sync *yet*, because one of them hasn't processed all its incoming messages) and an abnormal situation (the two services are not in sync because of a bug).
- Because of the previous point (can't differentiate between valid and invalid input), there's a lot of code that just does nothing in case of invalid input, instead of returning an error or printing a log message that could be helpful for detecting bugs.
- This design clearly doesn't scale in terms of complexity. If we continue adding services and inter-dependencies, we're going to increase the chances of lock contention and the complexity of the whole thing.
- It is often extremely valuable to split a module in two in order to reduce by half the complexity of both of them. However, splitting a module that contains Arcs and channels in two becomes quite difficult as you need to think about the order in which operations happen.
- As, I assume, the design was getting too complex, the `network::Context` trait for example has been used for multiple purposes when it shouldn't have been the case, leading to other entanglement issues.
- I'm generally trying to remove background threads in order to compile a light client for the browser (https://github.com/paritytech/substrate/issues/2416).
- The telemetry is particularly awkward because it is stored in a global variable, meaning that we store an Arc to half of the program's state in a global variable.

Here's what I suggest:

- The client, import queue, network, grandpa, informant, and so on, don't hold any Arc or any channel to any other service.
- The service holds ownership of everything.
- This is already the case more often than not, but as a first step replace storing `Arc<ConcreteType>` with `Arc<dyn Foo>` (`Foo` being a trait).
- Replace storing `Arc<dyn Foo>` as a field (`Foo` being a trait) with passing a `&mut dyn Foo` whenever a method is called, or with adding return values to methods that signal when some change has to be propagated to another part of the code. To give a concrete example, the `handle_received_network_message` method shouldn't do `self.import_queue.add(received_block)`. Instead it could either accept an additional `import_queue` parameter, or return a `Vec<Block>` with the list of blocks to put in the import queue.
- Crates don't directly depend on one another anymore. The only responsibility of each piece of code is to synchronously feed back information and actions to its direct owner.
- The only place in the code where we create threads is in service.
- The service crate is entirely responsible for concurrency. We don't want to entirely remove threads and passing messages around, but we can put all of it in the same Rust module. This single module can then be used to add debugging information (for example, add a timer every time we lock, or replace locks with our own timed lock), to decide of the tokio executor, or just in general to experiment. It would be very easy to start collecting metrics.
- Sub-services (eg. network, import queue, etc.) can spawn background tokio tasks, but they need a single-threaded fallback in the case where the tokio runtime isn't available (for #2416). See [an example of how to do that here](https://github.com/libp2p/rust-libp2p/blob/c2398adf67d2cadf257267d7d602865846a1ea0e/core/src/nodes/handled_node_tasks.rs#L323-L342).

While making this change in the entirety of Substrate would be an enormous task, I intend to work at least partially on this for the internals of the network crate. I would be glad if people helped, or at least kept this general objective in mind when designing new APIs.

Let me know about general feedback on this.

**Comment by gterzian** [comment]
I think a central component is not a bad idea, and I am opposed to the idea of it "owning everything" and the "synchronous feedback" parts. 

Two reasons:

1. You want to keep state isolated as much as possible, and not share it directly across threads, this holds whether the sharing is "decentralized", with various components owning references to each other like is often done now, or whether the sharing is done by a central component. 
2. You want components to run in parallel of each other, where possible. 

If I understand correctly what you describe, the central component would pass mutable reference around, presumably only one at a time, meaning no locks are required? However, it would also make it impossible to get any parallelism. It reads a little like you want to run a single-threaded event-loop and cycle through components, passing mutable references around to allow them to perform their required operations, similar to what is currently done in [`run_thread`](https://github.com/paritytech/substrate/blob/6bf7a175771863606561438c8fac2e624e3ececb/core/network/src/service.rs#L585).

Such a setup might be fine for a single component like network, where the intent is to merge pieces together onto a single thread or task, and I don't think it's a viable setup for Substrate as a whole, since it will make it very hard to do stuff in parallel of each other. 

Instead, I'd argue for the central component owning a bunch of `Box<Trait>`, which `Trait` being implemented internally without any shared state. The actual "state" and code manipulating the state, would then be isolated inside a thread, or task, and handle incoming messages. 

An example of such an approach is the [`ImportQueue`](https://github.com/paritytech/substrate/blob/6bf7a175771863606561438c8fac2e624e3ececb/core/consensus/common/src/import_queue.rs#L100). It's a trait, but it's implementation doesn't hold any state, instead it holds only a sender to a parallel component running the actual import queue logic. 

FInally, component should be allowed to own clones of such `Box<Trait>`, when direct communication between components would be more appropriate versus going through a central component acting as a kind of router. 

Also, I disagree with some of the description of problems with async operations. 

> Considering that everything is asynchronous, we can't really differentiate between a normal situation (the two services are not in sync yet, because one of them hasn't processed all its incoming messages) and an abnormal situation (the two services are not in sync because of a bug).

In order to differentiate between "normal" or "abnormal", one first needs to determine which operations can be async, and which need to be sync. Once that is done, it's easy to determine that a situation is abnormal, if you expect two components to be in sync with regards to a certain operation, and they're not, there's a bug somewhere and it needs to be fixed. 

The problem in my opinion is more one of properly determining your expectation with regards to how a component, or several components, will run. 

> It is extremely valuable to have APIs that return an error in case of bad input, but if you send your message over a channel you cannot have that feedback.

You can implement async feedback, in fact it is already done for example when the import-queue reports a peer for a bad block. 

If sync feedback is required, it can be be implemented as such, for example via a message containing a sender, and having the sending component block on awaiting the reply. 

> This single module can then be used to add debugging information (for example, add a timer every time we lock, or replace locks with our own timed lock)

This can be done even if the concurrency story is not completely centralized or synchronous. An example of such an approach is the [background-hang-monitor](https://github.com/servo/servo/tree/master/components/background_hang_monitor) in Servo, which uses async messaging to keep track of activity in various threads. 

I would be in favor in implementing something similar, and I don't think it's an argument to centralize everything, since it can be done in a decentralized way too. 

> To give a concrete example, the handle_received_network_message method shouldn't do self.import_queue.add(received_block). Instead it could either accept an additional import_queue parameter, or return a Vec<Block> with the list of blocks to put in the import queue.

I disagree with this design, since it implies all import-queue operations are sync, and/or a direct reference is held to the component storing the state of the import-queue. 

For example, if you wanted to make this `self.import_queue.add(received_block)` happen "in the background", you'd have to spin-up a thread, or use a thread-pool, and then pass the reference to `import-queue` in some way so that the work performed inside the method call would happen inside a different thread. So that mean you'd need to internally implement `import-queue` with locks, or if done with channels, you'd need to make the underlying communication sync, or else you won't be able to return the `Vec<Block> `. If the `import-queue` param is a mutable reference, and you avoid locks by having only one, you will not get any parallelism, and I don't see how you could to the import work in a separate thread from the network. 

If you mean to really just pass a `& dyn ImportQueue` or something to a method, and still implement things with async messaging inside the trait, I wonder what the point is of passing the import-queue as a parameter. Why not have the component that needs it hold a clone, at least it could potentially run in it's own thread or task and doesn't require the central component to pass it a reference to the import-queue on each method call.

**Comment by tomaka** [comment]
> If I understand correctly what you describe, the central component would pass mutable reference around, presumably only one at a time, meaning no locks are required?

I want to remove locks from the "sub-components" only. In other words, move the locks to the central component so that locking is easy to figure out.

> The actual "state" and code manipulating the state, would then be isolated inside a thread, or task, and handle incoming messages.

I want to make it an implementation detail

> An example of such an approach is the ImportQueue. It's a trait, but it's implementation doesn't hold any state, instead it holds only a sender to a parallel component running the actual import queue logic.

That's not something I want to change, except for the fact that we should be using tasks instead of threads.

What I want to change is the fact that we expose this background thread in the API of `ImportQueue`, namely with the `start()` and `stop()` methods. Having `start` and `stop` in the API basically screams "we have a background thread".

What I also want to change, is that the background task should send back the reputation changes and disconnections to the front task, which then only performs them, as opposed to having the background task perform said reputation changes and disconnects.
The objective is to make the moment where we perform updates on the network deterministic, instead of leaving to some background task the liberty of locking a mutex or sending messages whenever it wants.

> Once that is done, it's easy to determine that a situation is abnormal

Here's the issue: it is absolutely **not** easy.

Take the peerset for example. If you send an `Incoming`, then you receive a `Connected` with the same peer ID as the incoming connection, then you receive a `Dropped` with that same peer ID again, is this situation normal?
Supposing that the `Connected` was sent before the `Incoming` was received, did the peerset also send `Dropped` before receiving `Incoming`, in which case the incoming is still pending? Or did it send `Dropped` after receiving `Incoming`, in which case the incoming has been resolved?

Concurrency is absolutely **not** easy. We cannot even figure out what is valid or not for something as simple as the peerset manager.

> You can implement async feedback, in fact it is already done for example when the import-queue reports a peer for a bad block.

> If sync feedback is required, it can be be implemented as such, for example via a message containing a sender, and having the sending component block on awaiting the reply.

To give a concrete example, the import queue could look something like this:

```rust
impl ImportQueue {
    pub fn start_importing_block(&mut self, block: B) {
        self.to_bg_task.send(...);    // Send some message to the background task
        self.blocks_importing.insert(...);    // Also maintain some local state
    }

    pub fn is_importing(&self, block: &B) -> bool {
        self.blocks_importing.contains(block)  // Look up the local state
    }

    pub fn poll(&mut self, network_actions: &mut NetworkOut) {
         while let Ok(Async::Ready(Some(msg))) = self.from_bg.poll() {
             Msg::FinishedImporting(b) => { self.blocks_importing.remove(&b); }
             Msg::FailedBlockImport(block_author) => { network_actions.punish(block_author) }
         }
    }
}
```

You have synchronous feedback with `is_importing`, while the work actually happens in the background. The background task wouldn't have direct access to the network, so that the network locking strategy is performed by whatever calls `import_queue.poll()`.

The `ImportQueue` would become a non-racy state machine. Calling `is_importing` with the same block multiple times in a row would always return the same value, unless you call `poll` in between. I don't want to have racy APIs where `is_importing` would lock some mutex and look up what the background task is doing, as this is a common source of bugs.

**Comment by gterzian** [comment]
> You have synchronous feedback with is_importing, while the work actually happens in the background. The background task wouldn't have direct access to the network, so that the network locking strategy is performed by whatever calls import_queue.poll()

And what about the "local" state of `blocks_importing`? I think it's a great idea to make that state local, however if it is "local to the central component", it means it effectively is a contented resource to which whatever wants to use `start_importing_block` would have to wait for. 

It could work in specific situations,  maybe the import-queue, and I don't think it's a good general approach. Let's say two other "components" want to use the `start_importing_block` API. In the proposed design, it means they'd have to wait for their turn to use the API, and that would be orchestrated by the central component. 

The benefit of using the async  approach I am used to, is that if you have a `start_importing_block` API, two components running independently can use it without contention. The API is non-blocking, and the component using it can go straight back to whatever it was doing. 

Now off-course it creates opportunities for conflicting messages at the point of handling. If two components use `start_importing_block` in parallel, the import-queue could get conflicting "requests".
But there are ways of handling that in the overall design of the system. A simple one is for example not to share the `start_importing_block`API to more than one component if you don't want to receive conflicting messages. Another is to structure to logic so that conflicting messages don't matter. Even if you have only one component using `start_importing_block`, the non-blocking approach is still beneficiary since that single component can go straight back to work and doesn't need to "wait" for the underlying request in `start_importing_block` to be somehow handled. 

> The ImportQueue would become a non-racy state machine. Calling is_importing with the same block multiple times in a row would always return the same value, unless you call poll in between. 

Ok so it sounds like we're aiming for the same thing, but have somehow opposite views on how to do it. You can look at what I've done here: https://github.com/paritytech/substrate/commit/6f4c9955ec164b3be860096c7ab9107279b53366#diff-0c805d3ef65d16a3738f8270c6726d38

Basically, the import-queue previously was keeping track of things like the number of blocks it had in it's queue, and the "best number" and so on, and then made it available to `network/sync` via a [blocking call which returned the current `ImportQueueStatus`](https://github.com/paritytech/substrate/commit/6f4c9955ec164b3be860096c7ab9107279b53366#diff-fdedd07ae8ae881a0907be42dd80a840L402). 

What I did was removing that "state" from the background import-queue thread, and move it to `network/sync`. So it's exactly the design that you sketch, `sync` would update it's local state, and then send a message to the background thread. 

Then when a block is imported, it's `sync` that receives a message from the background-thread, and can update it's local state in response. That's essentially what you're doing in that `while let Ok(Async::Ready(Some(msg))) = self.from_bg.poll()` loop. 

So I absolutely agree that's how we should do things, however I don't see how this ties into having component not hold senders to each direction and so on. 

> Having start and stop in the API basically screams "we have a background thread".

Yes I think it's best to remove them, on the other hand you might need a way to tell the background-thread to stop, in cases were just dropping senders is not enough(maybe a central component for example always keeps one sender hence the stream/channel will never stop because all senders have been dropped).

Also just as a detail, the `start` API is currently only necessary because of the point where we create the `ImportQueue`, and the point at which we create the `Link`. The link is only created later, and is then passed to the `ImportQueue` via `start`. That doesn't in itself have something to do with starting the background worker, which starts earlier when the queue is created. So something like `link_ready` would be more appropriate.

**Comment by sjeohp** [comment]
A useful goal would be a `--num-threads` argument, requiring all tasks be spawned to the same fixed thread-pool and a polling strategy that functions on a single-thread.

**Comment by tomaka** [comment]
I'm going to close this as too difficult to achieve at this point.

**Comment by adamdossa** [comment]
@tomaka Any plans to resurrect this PR? i.e. making nodes process RPC calls multi-threaded?

**Comment by tomaka** [comment]
Processing RPC calls is already multithreaded as far as I know.
This issue was mostly about organizing the threads/tasks in a more organized way, mostly for better readability and debuggability.
